### PR TITLE
Custom fleet autoscaler resync interval

### LINKED ISSUE #1955: Configurable custom resync period for FleetAutoscaler
I would like to be able to change "resyncPeriod" of Fleet Autoscaler from 30 seconds.

**Is your feature request related to a problem? Please describe.**

### Testing

I want to shorten the time for Test using Fleet AutoScaler.

### Development environment

GameServer lifecycle is sometimes shorter than production.
In this case, there may not be enough GameServer and errors occur.
This problem can be avoided by inscreasing the buffer size, but this is not desirable due to its cost.

### Production environment

Our service is supposed to have a large number of accesses in an instant.
We want to have an option to shorten the "resyncPeriod" to be ready for spikes.

**Describe the solution you'd like**

I want to set the "resyncPeriod" from helm.
I suggest using "SharedInformerOption" and "customResync" mechanism.

**Describe alternatives you've considered**

### Testing
Wait 30 seconds or more for tests to complete

### Development environment
Inform the development teams that frequent allocation may cause errors

### Production environment 
Adjust the buffer size manually when access spikes are expected

### PR DESCRIPTION
**Pull request by jie-bao**
**What type of PR is this?**
/kind feature

**What this PR does / Why we need it**:
Add a new property `sync` to FleetAutoScaler for supporting custom sync interval.
See more details in: #1955 

**Which issue(s) this PR fixes**:

Closes #1955 

**Special notes for your reviewer**:
None

### CODE REVIEW

**Review by steven-supersolid — APPROVED** [review]
Just approving the package-lock.json change which is unrelated but at least syncs package-lock and package again :)

**Comment by markmandel** [comment]
Handholding testing, because change in CRDs. (I can't wait until this gets fixed), and restarting tests!

**Comment by jie-bao** [comment]
> Handholding testing, because change in CRDs. (I can't wait until this gets fixed), and restarting tests!

Sorry, I found a bug and fixed it. But the CI is blocked on the e2e-stable again. Is there a way I can work around it by myself?

**Comment by markmandel** [comment]
Huh, that's odd. I removed the helm install. I'll do it again.

I wish there was a way for you to fix things (or for it not to be a problem, but I'll handhold it through).

**Comment by markmandel** [comment]
```
--- FAIL: TestAutoscalerBasicFunctions (408.14s)
    fleetautoscaler_test.go:103: error waiting for fleet condition on fleet simple-fleet-qctp9
FAIL
time="2021-07-08 18:08:45.322" level=info msg="Namespace 1625767270 is deleted"
FAIL	agones.dev/agones/test/e2e	454.668s
FAIL
make[1]: *** [test-e2e-integration] Error 1
Makefile:273: recipe for target 'test-e2e-integration' failed
make[1]: Leaving directory '/workspace/build'
Makefile:265: recipe for target 'test-e2e' failed
make: *** [test-e2e] Error 2
```

Seems like a legitimate failure, but I'll run it again for you, just so we can be sure 👍🏻

**Comment by markmandel** [comment]
Nope, looks like a legit issue, same failure on `e2e-feature-gates`

```
--- FAIL: TestAutoscalerBasicFunctions (389.41s)
    fleetautoscaler_test.go:103: error waiting for fleet condition on fleet simple-fleet-mlq8d
FAIL
time="2021-07-08 18:58:56.791" level=info msg="Namespace 1625770241 is deleted"
FAIL	agones.dev/agones/test/e2e	495.298s
FAIL
Makefile:273: recipe for target 'test-e2e-integration' failed
```

**Comment by markmandel** [comment]
Kicking this through again. 

As an aside - if you would like some help running either individual or groups of the e2e tests locally (or any of the other tests), drop into #development on Slack - happy to help you resolve those issues!

**Comment by markmandel** [comment]
Now this is passing CI, I owe you a review. Apologies for the delay -- been swamped with a few things. Should open up tomorrow or friday.

**Review by markmandel — COMMENTED** [review]
I haven't yet done a deep dive into controller.go (first pass overall looks pretty good!), but wanted to get you some initial feedback on some little things since I know you've been waiting a bit.

File: pkg/apis/autoscaling/v1/fleetautoscaler.go
Comment (jie-bao) [reply pkg/apis/autoscaling/v1/fleetautoscaler.go:87]: I also don't have a strong opinion for using a pointer. I just kept the same struct as FleetAutoscalerPolicy. But FixedInterval is a struct, do you mean looking for 0 in the `FixedInterval.seconds`?

File: site/content/en/docs/Guides/feature-stages.md
Comment (jie-bao) [reply site/content/en/docs/Guides/feature-stages.md:34]: Updated

File: pkg/apis/autoscaling/v1/fleetautoscaler.go
Comment (jie-bao) [reply pkg/apis/autoscaling/v1/fleetautoscaler.go:60]: Sorry for missing that. Updated.

File: test/e2e/fleetautoscaler_test.go
Comment (jie-bao) [reply test/e2e/fleetautoscaler_test.go:437]: I add an e2e test with short sync interval and verify it works as expected when scaling up. But it's a bit difficult to verify the scale down time in e2e test. It depends on the fleet status so the FAS may either be updated in one interval or two interval. Do you have any suggestion on this? Or do you think any other situations worth test?

Comment (markmandel) [reply test/e2e/fleetautoscaler_test.go:437]: I'm not sure if I do 😞 just thought I'd ask the question. But lemme dig in today, see what I can see.

**Comment by markmandel** [comment]
Looks like a legit failure:
```
--- FAIL: TestAutoscalerWithDifferentInterval (103.28s)
    fleetautoscaler_test.go:371: 
        	Error Trace:	framework.go:319
        	            				fleetautoscaler_test.go:371
        	Error:      	Should be true
        	Test:       	TestAutoscalerWithDifferentInterval
    fleetautoscaler_test.go:385: 
        	Error Trace:	framework.go:319
        	            				fleetautoscaler_test.go:385
        	Error:      	Should be true
        	Test:       	TestAutoscalerWithDifferentInterval
```

I'm going to take this for a spin today as well and also do some deeper review 👍🏻

File: sdks/nodejs/package-lock.json
Comment (markmandel) [inline sdks/nodejs/package-lock.json:3]: Not sure why this increment is here? Removing the change would help with conflicts.

**Review by markmandel — COMMENTED** [review]
Digging deeper - we should include your docs here too: 
https://github.com/googleforgames/agones/blob/main/site/content/en/docs/Reference/fleetautoscaler.md

(Also will want to wrap it in `feature` shortcodes)

I think that covers the documentation changes.

File: pkg/apis/autoscaling/v1/fleetautoscaler.go
Comment (markmandel) [reply pkg/apis/autoscaling/v1/fleetautoscaler.go:88]: I think mu comment below on `ApplyDefaults` I think it covers my point better, and is probably less confusing 😄

File: pkg/fleetautoscalers/controller.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:330]: To be consistent, we can use `cache.MetaNamespaceKeyFunc(fas)` to get the key in the right format that it is used all the way through. Or maybe just pass the key through? You have it already in the parent functions it seems.

File: pkg/fleetautoscalers/controller.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:378]: @roberthbailey I'm 97.3% sure this correct to use (i.e. if the generation has changed, we know the spec has changed/it has been redeployed), but want to run it past you in case we got something wrong.

File: pkg/fleetautoscalers/controller.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:120]: In `Run(ctx context.Context, workers int)` when the `ctx` returns a value from `Done()`, we should send a message to all the termination channels to make sure all the fasThreads are terminated.

**Review by markmandel — COMMENTED** [review]
This is super close - just found a few things as I dug deeper into the controller today.

File: test/e2e/fleetautoscaler_test.go
Comment (markmandel) [reply test/e2e/fleetautoscaler_test.go:437]: Yeah, I looked at your e2e test -- it's so very possible for this to hit a race condition / just taking longer to process than expected.

🤔  I'm almost tempted to remove it. The only idea I could have is to test that something _didn't_ happen - but even then, depending on when the allocation occurs, the sync period could come right after.

So timing is probably better off being tested in unit tests -- actually that also gives me an idea. 💡

File: pkg/fleetautoscalers/controller.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:331]: We use https://pkg.go.dev/k8s.io/apimachinery/pkg/util/clock elsewhere in our code, so that we can control time during unit tests. We should do the same here, 

See: https://github.com/googleforgames/agones/blob/main/pkg/sdkserver/sdkserver.go#L83

Then we can write some unit tests to confirm things happen in the timeframes we expect. 👍🏻

File: sdks/nodejs/package-lock.json
Comment (jie-bao) [reply sdks/nodejs/package-lock.json:3]: I don't know why it's changed. Seems some automatic steps in the build process. But I change it back now.

File: pkg/apis/autoscaling/v1/fleetautoscaler.go
Comment (jie-bao) [reply pkg/apis/autoscaling/v1/fleetautoscaler.go:357]: I use the second option, thanks for your advice.

File: pkg/fleetautoscalers/controller.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go:330]: Updated to `cache.MetaNamespaceKeyFunc(fas)`. This can save a parameter and make sure the key won't be inconsistent with `fas` itself. Otherwise, I will have to call this in the AddEventHandler code, not many codes saved.

File: pkg/fleetautoscalers/controller.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go:120]: Good advice! Added it into `Run`.

**Comment by markmandel** [comment]
Looks like an error in linting / compilation:

```
docker run -t -e "TERM=xterm-256color" -e "GO111MODULE=on" --rm -v /workspace/build//.config/gcloud:/root/.config/gcloud -v ~/.kube/:/root/.kube -v ~/.config/helm:/root/.config/helm -v ~/.cache/helm:/root/.cache/helm -v /workspace:/go/src/agones.dev/agones -v /workspace/build//.gomod:/go/pkg/mod -v /workspace/build//.gocache:/root/.cache/go-build -w /go/src/agones.dev/agones  agones-build:59e55a859e bash -c \
	"golangci-lint run ./examples/... && golangci-lint run --timeout 15m ./..."
WARN [runner] Can't run linter goanalysis_metalinter: bodyclose: failed prerequisites: [buildssa@agones.dev/agones/test/e2e [agones.dev/agones/test/e2e.test]: analysis skipped: errors in package: [/go/src/agones.dev/agones/test/e2e/fleetautoscaler_test.go:426:20: cannot use &(autoscalingv1.FixedIntervalSync literal) (value of type *"agones.dev/agones/pkg/apis/autoscaling/v1".FixedIntervalSync) as "agones.dev/agones/pkg/apis/autoscaling/v1".FixedIntervalSync value in struct literal]] 
WARN [runner] Can't run linter unused: buildir: analysis skipped: errors in package: [/go/src/agones.dev/agones/test/e2e/fleetautoscaler_test.go:426:20: cannot use &(autoscalingv1.FixedIntervalSync literal) (value of type *"agones.dev/agones/pkg/apis/autoscaling/v1".FixedIntervalSync) as "agones.dev/agones/pkg/apis/autoscaling/v1".FixedIntervalSync value in struct literal] 
ERRO Running error: buildir: analysis skipped: errors in package: [/go/src/agones.dev/agones/test/e2e/fleetautoscaler_test.go:426:20: cannot use &(autoscalingv1.FixedIntervalSync literal) (value of type *"agones.dev/agones/pkg/apis/autoscaling/v1".FixedIntervalSync) as "agones.dev/agones/pkg/apis/autoscaling/v1".FixedIntervalSync value in struct literal] 
Makefile:362: recipe for target 'lint' failed
make: *** [lint] Error 3
```

You can run `make lint` if you want to reproduce locally.

File: pkg/fleetautoscalers/controller.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go:342]: I removed the e2e test and added a time interval check in the "create fas thread" unit tests.

File: test/e2e/fleetautoscaler_test.go
Comment (jie-bao) [reply test/e2e/fleetautoscaler_test.go:358]: I removed it because I feel it's not possible for ensuring the time condition in real environment. I added an unit test for it.

File: pkg/fleetautoscalers/controller.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller.go:203]:
```suggestion
// getFleetAutoscalerByKey get the Fleet Autoscaler by key
// a nil FleetAutoscaler returned indicates that an attempt to sync should not be retried, e.g.  if the FleetAutoscaler no longer exists.
```

Would suggest a small note here to explain what the convention here is if there is a nil FleetAutoscaler value.

File: pkg/fleetautoscalers/controller_test.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller_test.go:737]: Two thoughts:

1. Shouldn't there be something after the sleep to confirm that the function has been run twice? (maybe a counter in the update mock function?)
2. Mentioned [here](https://github.com/googleforgames/agones/pull/2171/files#r673558093), let's use [util/clock](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/clock) - as then we don't need to sleep, we can instead roll time forward without have to wait for real time to pass (which also provides less chance for flakey tests).

File: site/content/en/docs/Guides/feature-stages.md
Comment (markmandel) [inline site/content/en/docs/Guides/feature-stages.md:27]:
```suggestion
{{% feature expiryVersion="1.17.0" %}}
```

I just did this in another PR 😁

File: site/content/en/docs/Guides/feature-stages.md
Comment (markmandel) [inline site/content/en/docs/Guides/feature-stages.md:37]:
```suggestion
{{% feature publishVersion="1.17.0" %}}
```

File: site/content/en/docs/Reference/fleetautoscaler.md
Comment (markmandel) [inline site/content/en/docs/Reference/fleetautoscaler.md:12]:
```suggestion
{{% feature expiryVersion="1.17.0" %}}
```

File: site/content/en/docs/Reference/fleetautoscaler.md
Comment (markmandel) [inline site/content/en/docs/Reference/fleetautoscaler.md:43]:
```suggestion
{{% feature publishVersion="1.17.0" %}}
```

File: site/content/en/docs/Reference/fleetautoscaler.md
Comment (markmandel) [inline site/content/en/docs/Reference/fleetautoscaler.md:85]:
```suggestion
{{% feature expiryVersion="1.17.0" %}}
```

File: site/content/en/docs/Reference/fleetautoscaler.md
Comment (markmandel) [inline site/content/en/docs/Reference/fleetautoscaler.md:108]:
```suggestion
{{% feature publishVersion="1.17.0" %}}
```

File: pkg/fleetautoscalers/controller_test.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller_test.go:737]: I tried to add something like a counter in update mock function. It works when I run the single test. However it failed when I run `make test-go` due to race condition. So I have to check the interval inside the update mock function.

For your second suggestion, do you mean use the `clock.FakeClock` to skip waiting for real time? I tried but it looks like the update function will be only called once even I wait for a longer time. If I use `clock.RealClock` it looks like no difference with `time`. Is there other way to leverage the `util/clock`?

Comment (markmandel) [reply pkg/fleetautoscalers/controller_test.go:737]: > I tried to add something like a counter in update mock function. It works when I run the single test. However it failed when I run make test-go due to race condition. So I have to check the interval inside the update mock function.

Wrap the counter access and mutation in a mutex, that'll solve that problem 👍🏻 

> For your second suggestion, do you mean use the clock.FakeClock to skip waiting for real time?

Yep, exactly. You will need to use an instance of the `Clock` interface in the controller, and make sure to grab the ticker from it's `NewTicker` functionality. You can no longer access the `time` package directly, everything need to run through the `Clock` instance - then in the test you can switch out the `Clock` implementation with a `FakeClock` and control the flow of perceived time by the system that way.

Comment (jie-bao) [reply pkg/fleetautoscalers/controller_test.go]: Thanks for your guide! I updated the controller structure and related tests. Please take a look.

File: pkg/fleetautoscalers/controller_test.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller_test.go:761]: This is looking so much better 🥳 . Rather than a single sleep, which could be flaky, let's use an:
https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#Eventually

(let's see if I can get this right doing this within Github)

```suggestion
		assert.Eventually(t, func() bool {
		   return uint32(2) == fasUpdatedCount.Value()
		}, 10 * time.Second, 500 * time.Milliseconds)
```

Then if the processing takes a little longer than expected, we can retry a few times.

File: pkg/fleetautoscalers/controller_test.go
Comment (markmandel) [inline pkg/fleetautoscalers/controller_test.go:724]: So I just realised this - shouldn't this have a:

```
	utilruntime.FeatureTestMutex.Lock()
	defer utilruntime.FeatureTestMutex.Unlock()

	assert.NoError(t, utilruntime.ParseFeatures(string(utilruntime.FeatureCustomFasSyncInterval)+"=true"))
```

At the top? I'm actually not quite sure how it works otherwise? 🤔

File: pkg/fleetautoscalers/controller.go
Comment (roberthbailey) [inline pkg/fleetautoscalers/controller.go:389]: This should check (either instead of or in addition to this) whether err != nil.

File: pkg/fleetautoscalers/controller.go
Comment (roberthbailey) [inline pkg/fleetautoscalers/controller.go:347]: Does `fasThreads` need to be protected by a mutex to prevent simultaneous writes?

File: pkg/fleetautoscalers/controller.go
Comment (roberthbailey) [reply pkg/fleetautoscalers/controller.go:396]: Let me make sure I understand the design.

* Each fleet autoscaler gets a separate go routine which is always running in the background. Every `fas.Spec.Sync.FixedInterval.Seconds` it checks to see if the fleet should be scaled.
* Meanwhile, when this sync function is called, it may decide to remove and recreate the go routine for a fleet autoscaler because the generation number has changed. This will cause the fleet to be scaled immediately (not waiting for the next timer tick).

According to https://github.com/kubernetes/kubernetes/issues/67428 the generation for a CRD will only change when the spec changes (but **not** metadata like annotations or labels). 

So if the intent is to look only for spec changes, then this looks ok. But if changing annotations or labels should trigger an immediate scale check, then this is insufficient.

Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go:396]: That's exactly what I expect. We will ignore changes on annotations or labels.

File: pkg/fleetautoscalers/controller.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go:389]: Thanks for point out. Updated.

File: pkg/fleetautoscalers/controller_test.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller_test.go:724]: It looks like this is for preventing different feature impacts each other. Anyone, I added it.

File: pkg/fleetautoscalers/controller.go
Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go:347]: I'm not sure. Will one FAS be synced twice simultaneously? If so we may need a mutex.

Comment (jie-bao) [reply pkg/fleetautoscalers/controller.go]: Use sync.map instead of map

**Comment by jie-bao** [comment]
Hi, above build failed. But it doesn't look like an issue involved by this PR. Could anyone help retry it?

**Comment by roberthbailey** [comment]
Here's the error:

```
Starting e2e integration test!
go test -mod=vendor -race -parallel=32 agones.dev/agones/test/e2e --kubeconfig /root/.kube/config \
	--gameserver-image=gcr.io/agones-images/simple-game-server:0.3 \
	--pullsecret= \
	--feature-gates="" \
	--namespace=""
time="2021-08-02 04:51:31.777" level=info msg="Starting e2e test(s)" featureGates="CustomFasSyncInterval=false&Example=true&NodeExternalDNS=false&PlayerAllocationFilter=false&PlayerTracking=false&RollingUpdateOnReady=true&SDKWatchSendOnExecute=true&StateAllocationFilter=false" gameServerImage="gcr.io/agones-images/simple-game-server:0.3" namespace= perfOutputDir= pullSecret= stressTestLevel=0 version=
time="2021-08-02 04:51:31.849" level=info msg="cause for namespace deletion error" reason=NotFound
time="2021-08-02 04:51:31.849" level=error msg="failed to cleanup e2e namespaces" error="namespaces \"1627879575\" not found"
FAIL	agones.dev/agones/test/e2e	0.159s
FAIL
Makefile:273: recipe for target 'test-e2e-integration' failed
```

**Comment by roberthbailey** [comment]
I restarted the test.

**Comment by jie-bao** [comment]
Hi @markmandel , could you take a look of this PR again? I think it might be ready :)

File: site/content/en/docs/Reference/fleetautoscaler.md
Comment (markmandel) [inline site/content/en/docs/Reference/fleetautoscaler.md:74]:
```suggestion
  # [Stage:Alpha]
  # [FeatureFlag:CustomFasSyncInterval]
    # type of the sync. for now, only FixedInterval is available
```

Just aligning this with our [other examples](https://github.com/googleforgames/agones/blob/main/examples/gameserver.yaml#L81)

Ooh, that made me remember, can we add these changes in here too:
https://github.com/googleforgames/agones/blob/main/examples/fleet.yaml

Sorry about that, just realised.

**Review by markmandel — COMMENTED** [review]
Sorry about the delay! Just caught a minor documentation thing that i missed first time - but I think that is the last item!

File: site/content/en/docs/Reference/fleetautoscaler.md
Comment (jie-bao) [reply site/content/en/docs/Reference/fleetautoscaler.md:76]: 👍 Thank you for point out! Hope there's no other comments I missed.

**Comment by markmandel** [comment]
One final final thing (looks like I wasn't clear missed in my last comment)

The changes you made to the example yaml in site/content/en/docs/Reference/fleetautoscaler.md - add them into https://github.com/googleforgames/agones/blob/main/examples/fleet.yaml (it's basically the same thing content, just gives you two different ways to look at it) - that file is a full Fleet specification, so you can see all the options in a yaml file.

That's it. Once that's done, will approve and merge! 👍🏻  nice work!

**Comment by jie-bao** [comment]
> One final final thing (looks like I wasn't clear in my last comment)
> 
> The changes you made to the example yaml in site/content/en/docs/Reference/fleetautoscaler.md - add them into https://github.com/googleforgames/agones/blob/main/examples/fleet.yaml (it's basically the same thing content, just gives you two different ways to look at it) - that file is a full Fleet specification, so you can see all the options in a yaml file.
> 
> That's it. Once that's done, will approve and merge! 👍🏻 nice work!

I think you meant https://github.com/googleforgames/agones/blob/main/examples/fleetautoscaler.yaml since there's no FleetAutoScaler in fleet.yaml. I updated it.

**Review by markmandel — APPROVED** [review]
> I think you meant https://github.com/googleforgames/agones/blob/main/examples/fleetautoscaler.yaml since there's no FleetAutoScaler in fleet.yaml. I updated it.

Yes! That's the one! 🥳

**Pull request merged**
